### PR TITLE
MOSIP-15215 New Reg Proc API to return enrollment status

### PR DESCRIPTION
**Pull request by neosoftSatish**

### CODE REVIEW

File: registration-processor/init/registration-processor-registration-status-service/src/test/java/io/mosip/registration/processor/status/api/controller/RegistrationExternalStatusControllerTest.java [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+	public void searchSuccessTest() throws Exception {
+		doNothing().when(registrationStatusRequestValidator).validate((registrationStatusRequestDTO),
+				"mosip.registration.status");
+
+
+		this.mockMvc.perform(post("/externalstatus/search").accept(MediaType.APPLICATION_JSON_VALUE)
+				.cookie(new Cookie("Authorization", regStatusToJson)).contentType(MediaType.APPLICATION_JSON_VALUE)
+				.content(regStatusToJson.getBytes()).header("timestamp", "2019-05-07T05:13:55.704Z"))
+				.andExpect(status().isOk());
+	}
```
Comment (vishwa-vyom) [inline registration-processor/init/registration-processor-registration-status-service/src/test/java/io/mosip/registration/processor/status/api/controller/RegistrationExternalStatusControllerTest.java:196]: can you add the assert statements for response to check if for 1001 and 1002 it has the correct status and for 1003 it has the error message

**Review by vishwa-vyom — CHANGES_REQUESTED** [review]

File: registration-processor/init/registration-processor-registration-status-service/src/main/java/io/mosip/registration/processor/status/api/controller/RegistrationExternalStatusController.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+	@Autowired
+	private DigitalSignatureUtility digitalSignatureUtility;
+	
+	@Autowired
+	private Environment env;
+	
+	@Value("${registration.processor.signature.isEnabled}")
+	private Boolean isEnabled;
+	
+	private static final String REG_STATUS_SERVICE_ID = "mosip.registration.processor.registration.status.id";
```
[the remark this answers is not in the document]
Comment (neosoftSatish) [reply registration-processor/init/registration-processor-registration-status-service/src/main/java/io/mosip/registration/processor/status/api/controller/RegistrationExternalStatusController.java]: changed this property name...

File: registration-processor/init/registration-processor-registration-status-service/src/main/java/io/mosip/registration/processor/status/api/controller/RegistrationExternalStatusController.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+	 * @param registrationIds the registration ids
+	 * @return the response entity
+	 * @throws RegStatusAppException
+	 */
+	@PreAuthorize("hasAnyRole('REGISTRATION_ADMIN', 'REGISTRATION_OFFICER', 'REGISTRATION_SUPERVISOR','RESIDENT')")
+	@PostMapping(path = "/externalstatus/search", consumes = MediaType.APPLICATION_JSON_VALUE, produces = MediaType.APPLICATION_JSON_VALUE)
+	@ApiOperation(value = "Get the registration entity", response = RegistrationExternalStatusCode.class)
+	@ApiResponses(value = { @ApiResponse(code = 200, message = "Registration Entity successfully fetched"),
+			@ApiResponse(code = 400, message = "Unable to fetch the Registration Entity") })
+	public ResponseEntity<Object> search(
```
[the remark this answers is not in the document]
Comment (neosoftSatish) [reply registration-processor/init/registration-processor-registration-status-service/src/main/java/io/mosip/registration/processor/status/api/controller/RegistrationExternalStatusController.java:88]: changed this messages

File: registration-processor/init/registration-processor-registration-status-service/src/main/java/io/mosip/registration/processor/status/api/controller/RegistrationExternalStatusController.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+			@RequestBody(required = true) RegistrationStatusRequestDTO registrationStatusRequestDTO)
+			throws RegStatusAppException {
+
+		try {
+			registrationStatusRequestValidator.validate(registrationStatusRequestDTO,
```
[the remark this answers is not in the document]
Comment (neosoftSatish) [reply registration-processor/init/registration-processor-registration-status-service/src/main/java/io/mosip/registration/processor/status/api/controller/RegistrationExternalStatusController.java]: created new validator class...

File: registration-processor/init/registration-processor-registration-status-service/src/main/java/io/mosip/registration/processor/status/api/controller/RegistrationExternalStatusController.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+	 * @return the response entity
+	 * @throws RegStatusAppException
+	 */
+	@PreAuthorize("hasAnyRole('REGISTRATION_ADMIN', 'REGISTRATION_OFFICER', 'REGISTRATION_SUPERVISOR','RESIDENT')")
+	@PostMapping(path = "/externalstatus/search", consumes = MediaType.APPLICATION_JSON_VALUE, produces = MediaType.APPLICATION_JSON_VALUE)
+	@ApiOperation(value = "Get the registration entity", response = RegistrationExternalStatusCode.class)
+	@ApiResponses(value = { @ApiResponse(code = 200, message = "Registration Entity successfully fetched"),
+			@ApiResponse(code = 400, message = "Unable to fetch the Registration Entity") })
+	public ResponseEntity<Object> search(
+			@RequestBody(required = true) RegistrationStatusRequestDTO registrationStatusRequestDTO)
```
[the remark this answers is not in the document]
Comment (neosoftSatish) [reply registration-processor/init/registration-processor-registration-status-service/src/main/java/io/mosip/registration/processor/status/api/controller/RegistrationExternalStatusController.java]: created new requestDTO...

[the remark this answers is not in the document]
Comment (neosoftSatish) [reply registration-processor/init/registration-processor-registration-status-service/src/main/java/io/mosip/registration/processor/status/api/controller/RegistrationExternalStatusController.java:89]: created new requestDTO...

File: registration-processor/init/registration-processor-registration-status-service/src/main/java/io/mosip/registration/processor/status/api/controller/RegistrationExternalStatusController.java [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+					.getByIds(registrationStatusRequestDTO.getRequest());
+			
+			List<RegistrationStatusSubRequestDto> requestIdsNotAvailable = registrationStatusRequestDTO.getRequest()
+					.stream()
+					.filter(request -> registrations.stream().noneMatch(
+							registration -> registration.getRegistrationId().equals(request.getRegistrationId())))
+					.collect(Collectors.toList());
+			List<RegistrationStatusDto> registrationsList = syncRegistrationService.getByIds(requestIdsNotAvailable);
+			if (registrationsList != null && !registrationsList.isEmpty()) {
+				registrations.addAll(syncRegistrationService.getByIds(requestIdsNotAvailable));
```
[the remark this answers is not in the document]
Comment (neosoftSatish) [reply registration-processor/init/registration-processor-registration-status-service/src/main/java/io/mosip/registration/processor/status/api/controller/RegistrationExternalStatusController.java]: handled this...

File: registration-processor/init/registration-processor-registration-status-service/src/test/java/io/mosip/registration/processor/status/api/controller/RegistrationExternalStatusControllerTest.java (hunk 1 above)
Comment (neosoftSatish) [reply registration-processor/init/registration-processor-registration-status-service/src/test/java/io/mosip/registration/processor/status/api/controller/RegistrationExternalStatusControllerTest.java:196]: done.

File: registration-processor/init/registration-processor-registration-status-service/src/main/java/io/mosip/registration/processor/status/api/controller/RegistrationExternalStatusController.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (vishwa-vyom) [reply registration-processor/init/registration-processor-registration-status-service/src/main/java/io/mosip/registration/processor/status/api/controller/RegistrationExternalStatusController.java]: Unit test case class for validator class is missing

File: registration-processor/init/registration-processor-registration-status-service/src/main/java/io/mosip/registration/processor/status/api/controller/RegistrationExternalStatusController.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (vishwa-vyom) [reply registration-processor/init/registration-processor-registration-status-service/src/main/java/io/mosip/registration/processor/status/api/controller/RegistrationExternalStatusController.java]: Why separate response DTO is not created ?

File: registration-processor/init/registration-processor-registration-status-service/src/test/java/io/mosip/registration/processor/status/api/controller/RegistrationExternalStatusControllerTest.java (hunk 1 above)
Comment (vishwa-vyom) [reply registration-processor/init/registration-processor-registration-status-service/src/test/java/io/mosip/registration/processor/status/api/controller/RegistrationExternalStatusControllerTest.java:196]: 1002 is not added to asserts

Comment (neosoftSatish) [reply registration-processor/init/registration-processor-registration-status-service/src/test/java/io/mosip/registration/processor/status/api/controller/RegistrationExternalStatusControllerTest.java:196]: added 1002.

File: registration-processor/init/registration-processor-registration-status-service/src/main/java/io/mosip/registration/processor/status/api/controller/RegistrationExternalStatusController.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (neosoftSatish) [reply registration-processor/init/registration-processor-registration-status-service/src/main/java/io/mosip/registration/processor/status/api/controller/RegistrationExternalStatusController.java]: created separate dto for response

File: registration-processor/init/registration-processor-registration-status-service/src/main/java/io/mosip/registration/processor/status/api/controller/RegistrationExternalStatusController.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (neosoftSatish) [reply registration-processor/init/registration-processor-registration-status-service/src/main/java/io/mosip/registration/processor/status/api/controller/RegistrationExternalStatusController.java]: added unit test cases

**Review by vishwa-vyom — APPROVED** [review]

**Pull request merged**